pkg/server/endpoints/middleware_test.go

```diff
@@ -0,0 +1,188 @@
+package endpoints
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+	"github.com/sirupsen/logrus/hooks/test"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	"github.com/spiffe/spire/pkg/server/api"
+	"github.com/spiffe/spire/pkg/server/plugin/datastore"
+	"github.com/spiffe/spire/proto/spire-next/types"
+	"github.com/spiffe/spire/proto/spire/common"
+	"github.com/spiffe/spire/test/fakes/fakedatastore"
+	"github.com/spiffe/spire/test/spiretest"
+	"github.com/spiffe/spire/test/testca"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/codes"
+)
+
+func TestAuthorizedEntryFetcher(t *testing.T) {
+	ds := fakedatastore.New(t)
+
+	createEntry := func(entryIn *common.RegistrationEntry) *types.Entry {
+		resp, err := ds.CreateRegistrationEntry(context.Background(), &datastore.CreateRegistrationEntryRequest{
+			Entry: entryIn,
+		})
+		require.NoError(t, err)
+		entryOut, err := api.RegistrationEntryToProto(resp.Entry)
+		require.NoError(t, err)
+		return entryOut
+	}
+
+	setNodeSelectors := func(id spiffeid.ID, selectors []*common.Selector) {
+		_, err := ds.SetNodeSelectors(context.Background(), &datastore.SetNodeSelectorsRequest{
+			Selectors: &datastore.NodeSelectors{
+				SpiffeId:  id.String(),
+				Selectors: selectors,
+			},
+		})
+		require.NoError(t, err)
+	}
+
+	serverID := testTD.NewID("/spire/server")
+	agentID := testTD.NewID("/spire/agent")
+	nodeAliasID := testTD.NewID("/node-alias")
+	workload1ID := testTD.NewID("/workload1")
+	workload2ID := testTD.NewID("/workload2")
+
+	setNodeSelectors(agentID, []*common.Selector{
+		{Type: "node", Value: "value1"},
+		{Type: "node", Value: "value2"},
+	})
+
+	nodeAliasEntry := createEntry(&common.RegistrationEntry{
+		ParentId:  serverID.String(),
+		SpiffeId:  nodeAliasID.String(),
+		Selectors: []*common.Selector{{Type: "node", Value: "value1"}},
+	})
+
+	workload1Entry := createEntry(&common.RegistrationEntry{
+		ParentId:  agentID.String(),
+		SpiffeId:  workload1ID.String(),
+		Selectors: []*common.Selector{{Type: "workload", Value: "value1"}},
+	})
+
+	workload2Entry := createEntry(&common.RegistrationEntry{
+		ParentId:  agentID.String(),
+		SpiffeId:  workload2ID.String(),
+		Selectors: []*common.Selector{{Type: "workload", Value: "value2"}},
+	})
+
+	fetcher := AuthorizedEntryFetcher(ds)
+	entries, err := fetcher.FetchAuthorizedEntries(context.Background(), agentID)
+	require.NoError(t, err)
+	require.ElementsMatch(t, []*types.Entry{
```

Comment:
The datastore only have entries from this agent. It may be good to have an additional entry from a different agent so we make sure that only fetches the entries for the specified agent.
Considering the above comment, I'm wondering if having a table-driven test would be a better fit here.

Reply:
Ah, yes. I'll add another entry and assert it isn't returned.

As far as table-driven tests are concerned, I hesitate because there are really only two test cases here ( `regentryutil.FetchRegistrationEntries` succeeds or fails). The unit-tests for `regentryutil.FetchRegistrationEntries` provide all of the coverage for the crawl.